src/context/index.ts

```diff
@@ -0,0 +1,232 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { computed, IArrayWillSplice, intercept, observable, observe } from 'mobx'
+import { Log, Result } from 'sarif'
+import { commands, DiagnosticSeverity, ExtensionContext, languages, Memento, Range, Selection, TextDocument, ThemeColor, Uri, window, workspace } from 'vscode'
+import { mapDistinct, _Region, parseRegion } from '../shared'
+import '../shared/extension'
+import { Baser } from './Baser'
+import { loadLogs } from './loadLogs'
+import { Panel } from './Panel'
+
+declare module 'vscode' {
+	interface Diagnostic {
+		_result: Result
+	}
+}
+
+export const regionToSelection = (doc: TextDocument, region: _Region | undefined) => {
+	if (!region) return new Selection(0, 0, 0, 0) // TODO: Decide if empty regions should be pre-filtered.
+	return Array.isArray(region)
+		? (region.length === 4
+			? new Selection(...region)
+			: (() => {
+				const [byteOffset, byteLength] = region
+				const startColRaw = byteOffset % 16
+				const endColRaw = (byteOffset + byteLength) % 16
+				return new Selection(
+					Math.floor(byteOffset / 16),
+					10 + startColRaw + Math.floor(startColRaw / 2),
+					Math.floor((byteOffset + byteLength) / 16),
+					10 + endColRaw + Math.floor(endColRaw / 2),
+				)
+			})())
+		: (() => {
+			const line = doc.lineAt(region)
+			return new Selection(
+				line.range.start.line,
+				line.firstNonWhitespaceCharacterIndex,
+				line.range.end.line,
+				line.range.end.character,
+			)
+		})()
+	}
+
+export class Store {
+	static extensionPath: string | undefined
+	static globalState: Memento
+
+	@observable.shallow logs = [] as Log[]
+	@computed public get results() {
+		const runs = this.logs.map(log => log.runs).flat()
+		return runs.map(run => run.results).filter(run => run).flat() as Result[]
```

Comment:
>as Result[] [](start = 64, length = 11)

Is this "cast" necessary?

Reply:
Typescript normally infers this as `(Result | undefined)[]. It does not understand `filter(run => run)` removes the undefined values. Thus overriding the type to `Result[]`.

Reply:
Typescript thinks it is (Result | undefined)[]. It does not realize the undefined values have been filtered out. Thus the cast.

---
In reply to: [434591947](https://github.com/microsoft/sarif-vscode-extension/pull/279#discussion_r434591947) [](ancestors = 434591947)